plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/bugs/IncorrectDateTimeFormatInspection.java

```diff
@@ -0,0 +1,262 @@
+      @Override
+      public void visitMethodCallExpression(@NotNull PsiMethodCallExpression call) {
+        if (PATTERN_METHODS.test(call)) {
+          processExpression(call.getArgumentList().getExpressions()[0]);
+        }
+      }
+
+      private void processExpression(@NotNull PsiExpression expression) {
+        Object patternObject = ConstantExpressionUtil.computeCastTo(expression,
+                                                                    PsiType.getJavaLangString(expression.getManager(),
+                                                                                              expression.getResolveScope()));
+        if (!(patternObject instanceof String)) return;
+        String pattern = (String)patternObject;
+        List<Token> tokens = new ArrayList<>();
+        int optionalDepth = 0;
+        char[] array = pattern.toCharArray();
+        boolean inQuote = false;
+        int lastQuoteIndex = 0;
+        int length = 0;
+        for (int pos = 0; pos < array.length; pos++) {
+          char c = array[pos];
+
+          //process tokens
+          if (!inQuote && isLetter(c)) {
+            length++;
+            if (pos != array.length - 1 && c == array[pos + 1]) {
+              continue;
+            }
+            else {
+              tokens.add(new Token(pos - length + 1, array, length));
+              checkPadding(expression, array, pos - length + 1, pos + 1);
+            }
+          }
+          length = 0;
+
+
+          //check if it starts quote region or one quote
+          //case with one quote ('') is processed as quote region
+          if (c == '\'') {
+            inQuote = !inQuote;
+            lastQuoteIndex = pos;
+            continue;
+          }
+          //skip whole region in quote
+          if (inQuote) {
+            continue;
+          }
+
+          checkUnsupportedSymbolsProblem(expression, pos, c);
+
+          if (c == '[') {
+            optionalDepth++;
+          }
+          if (c == ']') {
+            optionalDepth--;
+          }
+        }
+
+        checkInQuoteProblem(expression, inQuote, lastQuoteIndex);
+        checkOptionalDepthProblem(expression, optionalDepth);
+
+        for (Token token : tokens) {
+          checkAvailableCount(expression, token, holder);
+        }
+      }
+
+      private void checkOptionalDepthProblem(@NotNull PsiExpression expression, int optionalDepth) {
+        if (optionalDepth < 0) {
+          holder.registerProblem(expression, null,
+                                 InspectionGadgetsBundle.message("inspection.incorrect.date.format.message.unpaired", ']'),
+                                 LocalQuickFix.EMPTY_ARRAY);
+        }
+      }
+
+      private void checkInQuoteProblem(@NotNull PsiExpression expression, boolean inQuote, int lastQuoteIndex) {
+        if (inQuote) {
+          TextRange range = ExpressionUtils.findStringLiteralRange(expression, lastQuoteIndex, lastQuoteIndex + 1);
+          if (range != null) {
```

Comment:
What if range is null? It's possible, e.g., if we are analysing the reference to a constant variable. Should we skip highlighting at all?

Reply:
Thank you for this comment. I fixed it and added one more test case.